crates/neon/src/types_impl/extract/buffer.rs

```diff
@@ -0,0 +1,240 @@
+        extract::{private, TryFromJs, TryIntoJs, TypeExpected},
+        JsArrayBuffer, JsBigInt64Array, JsBigUint64Array, JsBuffer, JsFloat32Array, JsFloat64Array,
+        JsInt16Array, JsInt32Array, JsInt8Array, JsTypedArray, JsUint16Array, JsUint32Array,
+        JsUint8Array, JsValue, Value,
+    },
+};
+
+/// Wrapper for converting between bytes and [`JsArrayBuffer`](JsArrayBuffer)
+pub struct ArrayBuffer<B>(pub B);
+
+impl<'cx, B> TryFromJs<'cx> for ArrayBuffer<B>
+where
+    for<'b> B: From<&'b [u8]>,
+{
+    type Error = TypeExpected<JsBuffer>;
+
+    fn try_from_js(
+        cx: &mut Cx<'cx>,
+        v: Handle<'cx, JsValue>,
+    ) -> NeonResult<Result<Self, Self::Error>> {
+        let v = match v.downcast::<JsArrayBuffer, _>(cx) {
+            Ok(v) => v,
+            Err(_) => return Ok(Err(Self::Error::new())),
+        };
+
+        Ok(Ok(ArrayBuffer(B::from(v.as_slice(cx)))))
+    }
+}
+
+impl<'cx, B> TryIntoJs<'cx> for ArrayBuffer<B>
+where
+    B: AsRef<[u8]>,
+{
+    type Value = JsArrayBuffer;
+
+    fn try_into_js(self, cx: &mut Cx<'cx>) -> JsResult<'cx, Self::Value> {
+        JsArrayBuffer::from_slice(cx, self.0.as_ref())
+    }
+}
+
+impl<B> private::Sealed for ArrayBuffer<B> {}
+
+/// Wrapper for converting between bytes and [`JsBuffer`](JsBuffer)
+pub struct Buffer<B>(pub B);
+
+impl<'cx, B> TryFromJs<'cx> for Buffer<B>
+where
+    for<'b> B: From<&'b [u8]>,
+{
+    type Error = TypeExpected<JsBuffer>;
+
+    fn try_from_js(
+        cx: &mut Cx<'cx>,
+        v: Handle<'cx, JsValue>,
+    ) -> NeonResult<Result<Self, Self::Error>> {
+        let v = match v.downcast::<JsBuffer, _>(cx) {
+            Ok(v) => v,
+            Err(_) => return Ok(Err(Self::Error::new())),
+        };
+
+        Ok(Ok(Buffer(B::from(v.as_slice(cx)))))
+    }
+}
+
+impl<'cx, B> TryIntoJs<'cx> for Buffer<B>
+where
+    B: AsRef<[u8]>,
+{
+    type Value = JsBuffer;
+
+    fn try_into_js(self, cx: &mut Cx<'cx>) -> JsResult<'cx, Self::Value> {
+        JsBuffer::from_slice(cx, self.0.as_ref())
+    }
+}
+
+impl<B> private::Sealed for Buffer<B> {}
+
+impl<'cx, T> TryIntoJs<'cx> for Vec<T>
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+    type Value = JsTypedArray<T>;
+
+    fn try_into_js(self, cx: &mut Cx<'cx>) -> JsResult<'cx, Self::Value> {
+        JsTypedArray::from_slice(cx, self.as_slice())
+    }
+}
+
+impl<'cx, T> TryFromJs<'cx> for Vec<T>
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+    type Error = TypeExpected<JsTypedArray<T>>;
+
+    fn try_from_js(
+        cx: &mut Cx<'cx>,
+        v: Handle<'cx, JsValue>,
+    ) -> NeonResult<Result<Self, Self::Error>> {
+        let v = match v.downcast::<JsTypedArray<T>, _>(cx) {
+            Ok(v) => v,
+            Err(_) => return Ok(Err(Self::Error::new())),
+        };
+
+        Ok(Ok(v.as_slice(cx).to_vec()))
+    }
+}
+
+impl<T> private::Sealed for Vec<T>
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+}
+
+impl<'cx, T> TryIntoJs<'cx> for Box<[T]>
```

Comment:
(Just discussed in person: let's drop the Box special cases here. Boxed slice is not as common as Vec and &[].)

Reply:
Removed.